### PR TITLE
[xbuild] Implement ItemDefinitionGroup

### PR DESCRIPTION
**Pull request by vestild**
I fixed some test, that failed under wine + .NET 3.5 and 4.0, some test that failed becouse of my locale (decimal comma).
Then adopt old Haakon Sporsheims patch with ItemDefinitionGroup.

### CODE REVIEW

**Comment by vestild** [comment]
@Therzok done!

File: mcs/class/Microsoft.Build.Engine/Test/Microsoft.Build.BuildEngine/ProjectTest.cs [hunk 1]
```diff
@@ -2231,6 +2259,7 @@ public void TestBuildSolutionProject ()
 			}
 		}
 
+#if net_4_0
```
Comment (akoeplinger) [inline mcs/class/Microsoft.Build.Engine/Test/Microsoft.Build.BuildEngine/ProjectTest.cs]: This needs to be all uppercase (`NET_4_0`), the #if constants are case-sensitive.

**Comment by akoeplinger** [comment]
approve

File: mcs/class/Microsoft.Build.Engine/Test/Microsoft.Build.BuildEngine/ProjectTest.cs [hunk 2]
```diff
@@ -1301,6 +1301,34 @@ public void TestBuildProjectFileXml8 ()
 		}
 
 		[Test]
+		public void TestBatchedMetadataEmptyBatch () {
```
Comment (akoeplinger) [inline mcs/class/Microsoft.Build.Engine/Test/Microsoft.Build.BuildEngine/ProjectTest.cs]: I tried this test without this PR and it passed there as well, so it's not really testing for the right problem.

**Comment by akoeplinger** [comment]
@vestild can you explain in more detail why the ToolsVersion=4.0 -> 3.5 change is needed?

**Comment by vestild** [comment]
> @vestild can you explain in more detail why the ToolsVersion=4.0 -> 3.5 change is needed?
@akoeplinger
I expect this tests pass under .NET 3.5 (not 4.0)
My first commit fix tests, that pass under .NET 4.5, but fails under 3.5

File: mcs/class/Microsoft.Build.Engine/Test/Microsoft.Build.BuildEngine/ProjectTest.cs (hunk 2 above)
Comment (vestild) [inline mcs/class/Microsoft.Build.Engine/Test/Microsoft.Build.BuildEngine/ProjectTest.cs]: What is PR?

Comment (akoeplinger) [inline mcs/class/Microsoft.Build.Engine/Test/Microsoft.Build.BuildEngine/ProjectTest.cs]: PR == pull request (this patch). The test passes even on Mono master, so it doesn't verify the fix.

Comment (vestild) [inline mcs/class/Microsoft.Build.Engine/Test/Microsoft.Build.BuildEngine/ProjectTest.cs]: It fixes the mistake in Haakon Sporsheim's realisation of ItemDefinitionGroup.
Pls see change of Microsoft.Build.BuildEngine/Project.cs in ffa32d6

**Comment by directhex** [comment]
build

**Comment by vestild** [comment]
I see 30 tests in  MonoTests.Microsoft.Build.BuildEngine fails.
But I can't reproduce it on my computer, so I can't understand a reason.
Is there an instruction to create environment like in build agent. Virtual machine image or something like.

**Comment by akoeplinger** [comment]
@vestild I see the same tests fail on my machine with your patch. Make sure you start from a clean working directory (e.g. run `git clean -xdf`) if you built Mono before.

**Comment by vestild** [comment]
Seems that my PR doesn't realy add ItemDefinitionGroup, it just doesn't work in fact. (But pass tests).
Does I need to close this PR until I fix it?

**Comment by vestild** [comment]
Hey, anybody!
Should i move to an other PRs commits fith (#27536 and #26395) and #19678 fixing?

**Pull request closed without merging**